src/main/java/software/amazon/smithy/lsp/project/ProjectManager.java

```diff
@@ -126,4 +155,51 @@ public Document getDocument(String uri) {
         }
         return project.getDocument(uri);
     }
+
+    /**
+     * Computes per-project file changes from the given file events.
+     *
+     * @param events The file events to compute per-project file changes from
+     * @return A map of project name to the corresponding project's changes
+     */
+    public Map<String, ProjectChanges> computeProjectChanges(List<FileEvent> events) {
+        // Note: we could eagerly compute these and store them, but project changes are relatively rare,
+        //  and doing it this way means we don't need to manage the state.
+        Map<String, PathMatcher> projectSmithyFileMatchers = new HashMap<>(nonDetachedProjects().size());
+        Map<String, PathMatcher> projectBuildFileMatchers = new HashMap<>(nonDetachedProjects().size());
+
+        Map<String, ProjectChanges> changes = new HashMap<>(nonDetachedProjects().size());
+
+        nonDetachedProjects().forEach((projectName, project) -> {
+            projectSmithyFileMatchers.put(projectName, ProjectFilePatterns.getSmithyFilesPathMatcher(project));
+            projectBuildFileMatchers.put(projectName, ProjectFilePatterns.getBuildFilesPathMatcher(project));
+
+            // Need these to be hash sets so they are mutable
+            changes.put(projectName, new ProjectChanges(new HashSet<>(), new HashSet<>(), new HashSet<>()));
```

Comment:
Nit: You can make an empty constructor for the record that does this.

Also, do we care at all about the order of project changes?

Reply:
Decided not to do this, because I like how this way is explicit about needing the sets to be mutable, whereas it isn't clear an empty ctor should init with mutable sets.